src/Tribe/Image/Uploader.php

```diff
@@ -0,0 +1,160 @@
+			$id = $existing ? $existing : $this->upload_file( $this->featured_image );
+		} elseif ( $post = get_post( $this->featured_image ) ) {
+			$id = $post && 'attachment' === $post->post_type ? $this->featured_image : false;
+		} else {
+			$id = false;
+		}
+
+		return $id;
+	}
+
+	/**
+	 * @param strin $file_url
+	 *
+	 * @return int
+	 */
+	protected function upload_file( $file_url ) {
+		if ( ! filter_var( $file_url, FILTER_VALIDATE_URL ) ) {
+			return false;
+		}
+
+		$contents = @file_get_contents( $file_url );
+		if ( false === $contents ) {
+			return false;
+		}
+
+		$upload = wp_upload_bits( basename( $file_url ), null, $contents );
+
+		if ( isset( $upload['error'] ) && $upload['error'] ) {
+			return false;
+		}
+
+		$type = '';
+		if ( ! empty( $upload['type'] ) ) {
+			$type = $upload['type'];
+		} else {
+			$mime = wp_check_filetype( $upload['file'] );
+			if ( $mime ) {
+				$type = $mime['type'];
+			}
+		}
+
+		$attachment = array(
+			'post_title'     => basename( $upload['file'] ),
+			'post_content'   => '',
+			'post_type'      => 'attachment',
+			'post_mime_type' => $type,
+			'guid'           => $upload['url'],
+		);
+
+		$id = wp_insert_attachment( $attachment, $upload['file'] );
+		wp_update_attachment_metadata( $id, wp_generate_attachment_metadata( $id, $upload['file'] ) );
+		update_post_meta( $id, '_tribe_importer_original_url', $file_url );
+
+		$this->maybe_init_attachment_guids_cache();
+		$this->maybe_init_attachment_original_urls_cache();
+
+		self::$attachment_guids_cache[ get_post( $id )->guid ] = $id;
+		self::$original_urls_cache[ $file_url ] = $id;
+
+		return $id;
+	}
+
+	protected function get_attachment_ID_from_url( $featured_image ) {
+		$this->maybe_init_attachment_guids_cache();
+		$this->maybe_init_attachment_original_urls_cache();
+
+		$guids_cache = self::$attachment_guids_cache;
+		$original_urls_cache = self::$original_urls_cache;
+		if ( isset( $guids_cache[ $featured_image ] ) ) {
+			return $guids_cache[ $featured_image ];
+		} elseif ( isset( $original_urls_cache[ $featured_image ] ) ) {
+			return $original_urls_cache[ $featured_image ];
+		}
+
+		return false;
+	}
+
+	protected function maybe_init_attachment_guids_cache() {
+		if ( false === self::$attachment_guids_cache ) {
+			/** @var \wpdb $wpdb */
+			global $wpdb;
+			$guids = $wpdb->get_results( "SELECT ID, guid FROM $wpdb->posts where post_type = 'attachment'" );
+
+			if ( $guids ) {
+				$keys = wp_list_pluck( $guids, 'guid' );
+				$values = wp_list_pluck( $guids, 'ID' );
+				self::$attachment_guids_cache = array_combine( $keys, $values );
+			} else {
+				self::$attachment_guids_cache = array();
+			}
+		}
+	}
+
+	protected function maybe_init_attachment_original_urls_cache() {
+		if ( false === self::$original_urls_cache ) {
+			/** @var \wpdb $wpdb */
+			global $wpdb;
+			$original_urls = $wpdb->get_results( "SELECT p.ID, pm.meta_value FROM $wpdb->posts p
+					JOIN $wpdb->postmeta pm
```

Comment:
This is nitpicky and perhaps just my personal preference - but if we split the query over many lines it seems like this sort of format aids readability (because then all lines of the query itself start at the same point):

``` php
$wpdb->get_results( "
	SELECT whatever
	FROM neat_tabulation
	WHERE tabulation = 'cool'
" );
```